src/uct/ib/rc/accel/gga_mlx5.c

```diff
@@ -187,6 +227,204 @@ uct_gga_mlx5_rkey_resolve(uct_ib_mlx5_md_t *md, uct_rkey_t rkey)
+    cleanup_ctx->super.qp   = self->super.tx.wq.super;
+    cleanup_ctx->super.reg  = self->super.tx.wq.reg;
+    cleanup_ctx->dma_opaque = self->dma_opaque;
+
+    uct_rc_mlx5_base_ep_cleanup(&self->super, &cleanup_ctx->super);
+}
+
+UCS_CLASS_DEFINE(uct_gga_mlx5_ep_t, uct_rc_mlx5_base_ep_t);
+UCS_CLASS_DEFINE_NEW_FUNC(uct_gga_mlx5_ep_t, uct_ep_t, const uct_ep_params_t *);
+UCS_CLASS_DEFINE_DELETE_FUNC(uct_gga_mlx5_ep_t, uct_ep_t);
+
+static ucs_status_t
+uct_gga_mlx5_ep_get_address(uct_ep_h tl_ep, uct_ep_addr_t *addr)
+{
+    uct_gga_mlx5_ep_t *ep               = ucs_derived_of(tl_ep,
+                                                         uct_gga_mlx5_ep_t);
+    uct_rc_mlx5_iface_common_t *iface   = ucs_derived_of(
+            tl_ep->iface, uct_rc_mlx5_iface_common_t);
+    uct_gga_mlx5_ep_address_t *gga_addr = (uct_gga_mlx5_ep_address_t*)addr;
+    uct_ib_md_t *md                     = uct_ib_iface_md(&iface->super.super);
+
+    uct_ib_pack_uint24(gga_addr->qp_num, ep->super.tx.wq.super.qp_num);
+    if (uct_rc_iface_flush_rkey_enabled(&iface->super)) {
+        gga_addr->flags      = UCT_GGA_MLX5_EP_ADDRESS_FLAG_FLUSH_RKEY;
+        gga_addr->flush_rkey = (md->flush_rkey >> 16);
+    } else {
+        gga_addr->flags      = 0;
+        gga_addr->flush_rkey = 0;
+    }
+
+    gga_addr->reserved = 0;
+    return UCS_OK;
+}
+
+static ucs_status_t
+uct_gga_mlx5_ep_connect_to_ep_v2(uct_ep_h tl_ep,
+                                 const uct_device_addr_t *device_addr,
+                                 const uct_ep_addr_t *ep_addr,
+                                 const uct_ep_connect_to_ep_params_t *params)
+{
+    uct_gga_mlx5_ep_t *ep             = ucs_derived_of(
+            tl_ep, uct_gga_mlx5_ep_t);
+    uct_rc_mlx5_iface_common_t *iface = ucs_derived_of(
+            tl_ep->iface, uct_rc_mlx5_iface_common_t);
+
+    const uct_gga_mlx5_dev_addr_t *gga_dev_addr  =
+            (uct_gga_mlx5_dev_addr_t*)device_addr;
+    const uct_ib_address_t *ib_addr              = &gga_dev_addr->ib_addr;
+    const uct_gga_mlx5_ep_address_t *gga_ep_addr =
+            (const uct_gga_mlx5_ep_address_t*)ep_addr;
+    uint32_t qp_num;
+    struct ibv_ah_attr ah_attr;
+    enum ibv_mtu path_mtu;
+    ucs_status_t status;
+
+    uct_ib_iface_fill_ah_attr_from_addr(&iface->super.super, ib_addr,
+                                        ep->super.super.path_index, &ah_attr,
+                                        &path_mtu);
+    ucs_assert(path_mtu != UCT_IB_ADDRESS_INVALID_PATH_MTU);
+
+    qp_num = uct_ib_unpack_uint24(gga_ep_addr->qp_num);
+    status = uct_rc_mlx5_iface_common_devx_connect_qp(
+            iface, &ep->super.tx.wq.super, qp_num, &ah_attr, path_mtu,
+            ep->super.super.path_index, iface->super.config.max_rd_atomic);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    ep->super.super.atomic_mr_offset = 0;
+    ep->super.super.flags           |= UCT_RC_EP_FLAG_CONNECTED;
+    ep->super.super.flush_rkey       =
```

Comment:
need to pass full 24 bit of flush_rkey in gga_addr, otherwise it's not reconstructed correctly here. see https://github.com/openucx/ucx/blob/d510358597f64bdd7c9f41a21e0dac34839d19ef/src/uct/ib/rc/accel/rc_mlx5_ep.c#L863